subsys: net: lib: rest_client: Add new REST client library (#5853)
nrfconnect/sdk-nrf: doc/nrf/libraries/networking/rest_client.rst

.. _lib_rest_client:

REST client
###########

.. contents::
   :local:
   :depth: 2

> The REST client library provides a way to perform REST request.

Reviewer: ```suggestion
The REST client library provides means for performing REST requests.
```

Overview
********

> REST client library creates a socket with TLS if requested and uses Zephyr's HTTP client library to send HTTP request and receive HTTP response.
> The library is used by populating :c:struct:`rest_client_req_resp_context` structure passed to :c:func:`rest_client_request` function
> and the same structure will contain the response data.

Reviewer: ```suggestion
This library creates a socket with TLS when requested.
It uses Zephyr's HTTP client library to send HTTP requests and receive HTTP responses.
When using this library, the :c:struct:`rest_client_req_resp_context` structure is populated and passed to the :c:func:`rest_client_request` function.
The same structure will contain the response data.
```

Configuration
*************

> To enable the REST client library, enable the :kconfig:`CONFIG_REST_CLIENT` Kconfig option.

Reviewer: ```suggestion
To use the REST client library, enable the :kconfig:`CONFIG_REST_CLIENT` Kconfig option.
```

> Following are the options that can be configured to alter the behavior of the library:

Reviewer: ```suggestion
You can configure the following options to adjust the behavior of the library:
```

.. _lib_rest_client:

REST client
###########

.. contents::
   :local:
   :depth: 2

The REST client library provides a way to perform REST request.

Overview
********

REST client library creates a socket with TLS if requested and uses Zephyr's HTTP client library to send HTTP request and receive HTTP response.
The library is used by populating :c:struct:`rest_client_req_resp_context` structure passed to :c:func:`rest_client_request` function
and the same structure will contain the response data.

Configuration
*************

To enable the REST client library, enable the :kconfig:`CONFIG_REST_CLIENT` Kconfig option.

Following are the options that can be configured to alter the behavior of the library:

*  :kconfig:`CONFIG_REST_CLIENT_REQUEST_TIMEOUT`
*  :kconfig:`CONFIG_REST_CLIENT_SCKT_SEND_TIMEOUT`
*  :kconfig:`CONFIG_REST_CLIENT_SCKT_RECV_TIMEOUT`
*  :kconfig:`CONFIG_REST_CLIENT_SCKT_TLS_SESSION_CACHE_IN_USE`

Limitations
***********

> *  Executing REST request is a blocking operation.

Reviewer: This is a bit unclear, at least to me. Can you explain a bit more? And if this is the only limitation, you don't need the bullet point.

.. _lib_rest_client:

REST client
###########

.. contents::
   :local:
   :depth: 2

The REST client library provides means for performing REST requests.

Overview
********

This library creates a socket with TLS when requested.
It uses Zephyr's HTTP client library to send HTTP requests and receive HTTP responses.
When using this library, the :c:struct:`rest_client_req_resp_context` structure is populated and passed to the :c:func:`rest_client_request` function.
The same structure will contain the response data.

Configuration
*************

To use the REST client library, enable the :kconfig:`CONFIG_REST_CLIENT` Kconfig option.

You can configure the following options to adjust the behavior of the library:

*  :kconfig:`CONFIG_REST_CLIENT_REQUEST_TIMEOUT`
*  :kconfig:`CONFIG_REST_CLIENT_SCKT_SEND_TIMEOUT`
*  :kconfig:`CONFIG_REST_CLIENT_SCKT_RECV_TIMEOUT`
*  :kconfig:`CONFIG_REST_CLIENT_SCKT_TLS_SESSION_CACHE_IN_USE`

Limitations
***********

> *  Executing REST request is a blocking operation.

Reviewer: ```suggestion
Executing a REST request is a blocking operation.
The application thread is blocked until until the request has completed.
```

Author: I will add the following:
`The calling thread is blocked until the request has completed.`
